vulcan/lib/commands.rb

```diff
@@ -18,6 +19,174 @@ def setup(config)
     end
   end
 
+  class Workflow
+    include Etna::CommandExecutor
+
+    class Run
+      def execute(project_name, workflow_name)
+        @project_name = project_name
+        @workflow_name = workflow_name
+
+        while true
+          orchestration.run_until_done!(storage)
+          find_all_output_ui_queries.each do |query_name, source, path|
```

Comment:
So, this was just one novel interpretation of how to handle UI displaying outputs.  Essentially,

1.  primary outputs are mainly used for composing workflows together.  primary outputs of one workflow can be used as step outputs in another workflow in CWL.  Unfortunately that means that primary outputs don't get their own run specification and cannot explicitly define a ui query.  In this case, I assign primary outputs a default ui_query based on the type, although that does bring up the original point; does the frontend need to map types to outputs?  Maybe, but I'd like to avoid that.  Maybe we don't show primary outputs without an explicit cell anyways, thus allowing a workflow to both be a data pipeline as well as an interactive workflow -- the interactive inputs and outputs only get triggered, essentially, on a client that specifically handles them.  :shrug:  There are several options, but this one is simple.

2.  explicit steps with a run of `ui-queries/XXX` that themselves have no outputs, are considered UI components that display their inputs.  I encode that here as simply using `send` to dynamically dispatch output based on the ui-queries name.